docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,584 @@
+Additionally, we will update the `CacheWrap` and `CacheWrapper` interfaces to enable listening in the caching layer.
+
+```go
+type MultiStore interface {
+	...
+
+	// ListeningEnabled returns if listening is enabled for the KVStore belonging the provided StoreKey
+	ListeningEnabled(key StoreKey) bool
+
+	// SetListeners sets the WriteListeners for the KVStore belonging to the provided StoreKey
+	// It appends the listeners to a current set, if one already exists
+	SetListeners(key StoreKey, listeners []WriteListener)
+}
+```
+
+```go
+type CacheWrap interface {
+	...
+
+	// CacheWrapWithListeners recursively wraps again with listening enabled
+	CacheWrapWithListeners(listeners []WriteListener) CacheWrap
+}
+
+type CacheWrapper interface {
+	...
+
+	// CacheWrapWithListeners recursively wraps again with listening enabled
+	CacheWrapWithListeners(listeners []WriteListener) CacheWrap
+}
+```
+
+### MultiStore implementation updates
+
+We will modify all of the `Store` and `MultiStore` implementations to satisfy these new interfaces, and adjust the `rootmulti` `GetKVStore` method
+to wrap the returned `KVStore` with a `listenkv.Store` if listening is turned on for that `Store`.
+
+```go
+func (rs *Store) GetKVStore(key types.StoreKey) types.KVStore {
+	store := rs.stores[key].(types.KVStore)
+
+	if rs.TracingEnabled() {
+		store = tracekv.NewStore(store, rs.traceWriter, rs.traceContext)
+	}
+	if rs.ListeningEnabled(key) {
+		store = listenkv.NewStore(store, rs.listeners[key])
+	}
+
+	return store
+}
+```
+
+We will also adjust the `cachemulti` constructor methods and the `rootmulti` `CacheMultiStore` method to forward the listeners
+to and enable listening in the cache layer.
+
+```go
+func (rs *Store) CacheMultiStore() types.CacheMultiStore {
+	stores := make(map[types.StoreKey]types.CacheWrapper)
+	for k, v := range rs.stores {
+		stores[k] = v
+	}
+	return cachemulti.NewStore(rs.db, stores, rs.keysByName, rs.traceWriter, rs.traceContext, rs.listeners)
+}
+```
+
+### Exposing the data
+
+We will introduce a new `StreamingService` interface for exposing `WriteListener` data streams to external consumers.
+
+```go
+// StreamingService interface for registering WriteListeners with the BaseApp and updating the service with the ABCI context
+type StreamingService interface {
+	Listeners() map[sdk.StoreKey][]storeTypes.WriteListener // returns the streaming service's listeners for the BaseApp to register
+	BeginBlockReq(req abci.RequestBeginBlock) // update the streaming service with the latest RequestBeginBlock message
+	BeginBlockResres abci.ResponseBeginBlock) // update the steaming service with the latest ResponseBeginBlock message
+	EndBlockReq(req abci.RequestEndBlock) // update the steaming service with the latest RequestEndBlock message
+	EndBlockRes(res abci.ResponseEndBlock) // update the steaming service with the latest ResponseEndBlock message
```

Comment:
Is this out of scope of this ADR, which as I understand only adds listeners on the store level?

Also, these listeners seems more appropriate on Tendermint.

Reply:
IMO it is in scope. This relates back to Aaron's latest comments, and to how we want to maintain the relation of state changes to the ABCI messages/block context that caused them. This is important for downstream IPLD-ization, we need to maintain the hash-linked nature of the data or be able to reconstitute it.

Reply:
Yeah otherwise it's hard to know what block has been cached. They're just raw kv pairs with no time info. So even if we don't care about proofs we still need to commit the data to Postgres, etc transactionally.